src/storage/v2/disk/storage.cpp

```diff
@@ -0,0 +1,1601 @@
+  }
+  if (status.ok()) {
+    spdlog::debug("rocksdb: Saved edge with key {} and ts {}", serializedEdgeKey, *commit_timestamp_);
+  } else if (status.IsBusy()) {
+    spdlog::error("rocksdb: Edge with key {} and ts {} was changed and committed in another transaction",
+                  serializedEdgeKey, *commit_timestamp_);
+    return false;
+  } else {
+    spdlog::error("rocksdb: Failed to save edge with key {} and ts {}", serializedEdgeKey, *commit_timestamp_);
+    return false;
+  }
+  return true;
+}
+
+bool DiskStorage::DiskAccessor::DeleteVertexFromDisk(const std::string &vertex) {
+  auto *disk_storage = static_cast<DiskStorage *>(storage_);
+  auto status = disk_transaction_->Delete(disk_storage->kvstore_->vertex_chandle, vertex);
+  if (status.ok()) {
+    spdlog::debug("rocksdb: Deleted vertex with key {}", vertex);
+  } else if (status.IsBusy()) {
+    spdlog::error("rocksdb: Vertex with key {} was changed and committed in another transaction", vertex);
+    return false;
+  } else {
+    spdlog::error("rocksdb: Failed to delete vertex with key {}", vertex);
+    return false;
+  }
+  return true;
+}
+
+bool DiskStorage::DiskAccessor::DeleteEdgeFromDisk(const std::string &edge) {
+  auto *disk_storage = static_cast<DiskStorage *>(storage_);
+  auto status = disk_transaction_->Delete(disk_storage->kvstore_->edge_chandle, edge);
+  if (status.ok()) {
+    spdlog::debug("rocksdb: Deleted edge with key {}", edge);
+  } else if (status.IsBusy()) {
+    spdlog::error("rocksdb: Edge with key {} was changed and committed in another transaction", edge);
+    return false;
+  } else {
+    spdlog::error("rocksdb: Failed to delete edge with key {}", edge);
+    return false;
+  }
+  return true;
+}
+
+/// TODO: add the comment about the storage lock you should take for correctly operating on this method
+[[nodiscard]] utils::BasicResult<StorageDataManipulationError, void>
+DiskStorage::DiskAccessor::CheckConstraintsAndFlushMainMemoryCache() {
+  auto vertex_acc = vertices_.access();
+  uint64_t num_ser_edges = 0;
+
+  std::vector<std::vector<PropertyValue>> unique_storage;
+  auto *disk_unique_constraints =
+      static_cast<DiskUniqueConstraints *>(storage_->constraints_.unique_constraints_.get());
+  auto *disk_label_index = static_cast<DiskLabelIndex *>(storage_->indices_.label_index_.get());
+  auto *disk_label_property_index =
+      static_cast<DiskLabelPropertyIndex *>(storage_->indices_.label_property_index_.get());
+
+  /// TODO: andi I don't like that std::optional is used for checking errors but that's how it was before, refactor!
+  for (Vertex &vertex : vertex_acc) {
+    /// TODO: refactor this check, it is unreadable
+    if (auto existence_constraint_validation_result = storage_->constraints_.existence_constraints_->Validate(vertex);
+        existence_constraint_validation_result.has_value()) {
+      return StorageDataManipulationError{existence_constraint_validation_result.value()};
```

Comment:
Maybe I don't understand something, but here we are checking vertices one by one, and if there are not constraint violations we are storing them to rocks db. But what if constraint violation happens later on some vertex and we stored already a lot of them?